ingestion-beam/src/main/java/com/mozilla/telemetry/schema/SchemaStoreSingletonFactory.java

```diff
@@ -0,0 +1,55 @@
+package com.mozilla.telemetry.schema;
+
+import com.mozilla.telemetry.ingestion.core.schema.JSONSchemaStore;
+import com.mozilla.telemetry.ingestion.core.schema.PipelineMetadataStore;
+import com.mozilla.telemetry.util.BeamFileInputStream;
+
+/**
+ * Factory class for getting singleton instances of SchemaStores.
+ *
+ * <p>Beam can run up to 500 threads per worker
+ * (see https://cloud.google.com/dataflow/docs/guides/troubleshoot-oom#dofn)
+ * By using singleton instances of SchemaStores we can speed up worker startup a bit and
+ * decrease memory utilization.
+ *
+ * <p>`schemasLocation` parameter in the factory methods here is used only during object creation.
+ * This is fine here since the value is not changing during runtime of the job.
+ */
+public class SchemaStoreSingletonFactory {
+
+  private static volatile JSONSchemaStore jsonSchemaStore;
+  private static volatile PipelineMetadataStore pipelineMetadataStore;
+
+  private SchemaStoreSingletonFactory() {
+  }
+
+  /**
+   * Returns a singleton instance of {@link JSONSchemaStore}.
+   */
+  public static JSONSchemaStore getJsonSchemaStore(String schemasLocation) {
+    if (jsonSchemaStore == null) {
```

Comment:
Double-checked locking is used here because this will be called in `setup()` from multiple `DoFn` instances.